File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.cpp


//  struct ListNode {
//      int val;
//      ListNode *next;
//      ListNode() : val(0), next(nullptr) {}
//      ListNode(int x) : val(x), next(nullptr) {}
//      ListNode(int x, ListNode *next) : val(x), next(next) {}
//  };


class Solution {
public:

    ListNode* vectorToList(vector<int> &v) {
        if(v.size() == 0) return nullptr;
        ListNode *head = new ListNode(v[0]);
        ListNode *tail = head;
        for(int i = 1 ; i < v.size() ; i++) {
            tail -> next =new ListNode(v[i]);
            tail = tail->next;
        }
        return head;
    }

    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        vector<int>v = {};
        ListNode* temp1 = list1;
        while (temp1 != NULL) {
            v.push_back(temp1->val);
            temp1 = temp1->next;
        }
        ListNode* temp2 = list2;
        while (temp2 != NULL) {
            v.push_back(temp2->val);
            temp2 = temp2->next;
        }

        sort(v.begin(),v.end());

        return vectorToList(v);
        

    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        // finds the middle of the list(slow)
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode second = slow.next;  //4
        slow.next = null;
        ListNode prev = null;
        // 1 2 3 4 5 6 
        while(second != null) {
            ListNode temp = second.next; //5 
            second.next = prev; //null

            prev = second; //node 4
            second = temp; // 5
        }

        ListNode first = head;
        second = prev;
        // reverse order : 1 2 3 6 5 4
        while(second != null) {
            ListNode temp1 = first.next; //2
            ListNode temp2 = second.next; //5

            first.next = second; //6
            second.next = temp1; // 2

            first = temp1; //2
            second = temp2; // 5

        }

    }
}
